lib/sshutils/sftp/sftp.go

```diff
@@ -323,94 +320,112 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
 	}
 
 	if c.getHomeDir == nil {
-		c.getHomeDir = func() (string, error) {
-			return getRemoteHomeDir(sshClient)
-		}
+		c.getHomeDir = getLocalHomeDir
 	}
 
 	return trace.Wrap(c.expandPaths(srcOK, dstOK))
 }
 
 func (c *Config) expandPaths(srcIsRemote, dstIsRemote bool) (err error) {
-	srcHomeRetriever := getLocalHomeDir
-	if srcIsRemote {
-		srcHomeRetriever = c.getHomeDir
-	}
 	for i, srcPath := range c.srcPaths {
-		c.srcPaths[i], err = expandPath(srcPath, srcHomeRetriever)
+		c.srcPaths[i], err = expandPath(srcPath, srcIsRemote, c.getHomeDir)
 		if err != nil {
-			return trace.Wrap(err)
+			return trace.Wrap(err, "error expanding %q", srcPath)
 		}
 	}
 
-	dstHomeRetriever := getLocalHomeDir
-	if dstIsRemote {
-		dstHomeRetriever = c.getHomeDir
-	}
-	c.dstPath, err = expandPath(c.dstPath, dstHomeRetriever)
-
-	return trace.Wrap(err)
+	c.dstPath, err = expandPath(c.dstPath, dstIsRemote, c.getHomeDir)
+	return trace.Wrap(err, "error expanding %q", c.dstPath)
 }
 
-func getLocalHomeDir() (string, error) {
-	u, err := user.Current()
-	if err != nil {
-		return "", trace.Wrap(err)
+func expandPath(pathStr string, isRemotePath bool, getHomeDir homeDirRetriever) (string, error) {
+	pfxLen := homeDirPrefixLen(pathStr)
+	if pfxLen == -1 {
+		return pathStr, nil
 	}
-	return u.HomeDir, nil
-}
 
-func expandPath(pathStr string, getHomeDir homeDirRetriever) (string, error) {
-	if !needsExpansion(pathStr) {
-		return pathStr, nil
+	// if the path is local, expand the path to be rooted at this user's
+	// home dir
+	if !isRemotePath {
```

Comment:
A general question about this block. Typically the shell will expand `~`. Are we expanding `~` here for local paths incase a shell was not used to invoke `tsh scp`?

I think our guiding light here should be OpenSSH interoperability. Will OpenSSH ever expand `~` for local paths?

Reply:
That's a good point, OpenSSH `scp` does not expand local paths so I'll remove that code